### PR TITLE
store role assignments in the database

### PR DESCRIPTION
**Pull request by davepacheco**
This depends on #512.  This change moves role assignments to the database.  This should remove almost all the hardcoding from the policy (aside from the policy itself).  (The remaining bit of hardcoded behavior is that the special "db-init" user is hardcoded to have the "init" role on the database.  This is necessary because this user uses this role to load the rest of the built-in users and roles.)

Everything here is still "built-in users" with "built-in roles".  I'm reluctant to generalize further without knowing what IdP-based users or custom roles will look like.  I think we may want these to be in separate tables, which is why these tables all have "builtin" in the name.  That said, I'm leaning towards replacing "built-in users" with service accounts, which we already know we need -- that way there are fewer distinct kinds of actor.

There are a few side effects of this change:

* This change involved some substantial surgery on the authz subsystem.  I also split up oso_types into a bunch of separate files and added some more detailed docs on how authz works.
* The `authorize()` functions are now async.  I think this is unavoidable.  (It's also not a big deal.)  In order to authorize you, we have to look up roles in the database.  We prefetch the roles, but we can't do that until we know what resource you're authorized on, which we don't know until `authorize()`.
* Any automated tests that use the datastore now require not just that the datastore be created, but also that the built-in users, roles, and role assignments have been loaded into it.  Normally, this happens when Nexus starts up.  So integration tests are all set.  However, the datastore tests themselves now need to not just create a datastore but load all the built-in stuff.  This might make them a bit slower, but they were already starting a whole CockroachDB instance and this is just another few transactions.
* Similarly, the authz tests themselves now depend on a datastore, since doing authz now requires the database.  These were previously very quick tests that just loaded the Oso policy and made checks against it -- now they have to set up a whole database and load the built-in stuff.  Still, that should be pretty quick.

Status: I want to take a serious edit pass on the documentation in the authz crate, but otherwise I think this is ready for a look.

### CODE REVIEW

**Comment by davepacheco** [comment]
Marking "draft" because this depends on #512.

File: nexus/src/authz/api_resources.rs [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+impl Fleet {
+    /// Returns an authz resource representing a child Organization
+    pub fn organization(&self, organization_id: Uuid) -> Organization {
+        Organization { organization_id }
+    }
+
+    /// Returns an authz resource representing some other kind of child (e.g.,
+    /// a built-in user, built-in role, etc. -- but _not_ an Organization)
+    ///
+    /// Aside from Organizations (which you create with [`Fleet::organization()`
```
Comment (bnaecker) [inline nexus/src/authz/api_resources.rs:55]: nit: missing a closing `]` here.

File: nexus/src/authz/api_resources.rs [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+impl oso::PolarClass for FleetChild {
+    fn get_polar_class_builder() -> oso::ClassBuilder<Self> {
+        oso::Class::builder()
+            .with_equality_check()
+            .add_method(
+                "has_role",
+                /* Roles are not supported on FleetChilds today. */
+                |_: &FleetChild, _: AuthenticatedActor, _: String| false,
+            )
+            .add_attribute_getter("fleet", |_: &FleetChild| Fleet {})
```
Comment (bnaecker) [inline nexus/src/authz/api_resources.rs:112]: I don't know that it matters, but I expected this to be the singleton `FLEET`, rather than a new object.

File: nexus/src/authz/api_resources.rs [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+    }
+}
+
+impl AuthzApiResource for FleetChild {
+    fn db_resource(&self) -> Option<(ResourceType, Uuid)> {
+        None
+    }
+
+    fn parent(&self) -> Option<Box<dyn AuthzResource>> {
+        Some(Box::new(Fleet {}))
```
Comment (bnaecker) [inline nexus/src/authz/api_resources.rs:122]: Same thing here. Any equality check would of course still pass, but that won't be true if we change the `Fleet` type to include state. It might also be good documentation to identify this as `FLEET`. Again, assuming that's actually what we want, I'm still getting my bearings in this code.

File: nexus/src/authz/api_resources.rs [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+
+    /// Returns an authz resource representing some other kind of child (e.g.,
+    /// a built-in user, built-in role, etc. -- but _not_ an Organization)
+    ///
+    /// Aside from Organizations (which you create with [`Fleet::organization()`
+    /// instead), all instances of all types of Fleet children are treated
+    /// interchangeably by the authz subsystem.  That's because we do not
+    /// currently support assigning roles to these resources, so all that
+    /// matters for authz is that they are a child of the Fleet.
+    pub fn child_generic(&self) -> FleetChild {
```
Comment (bnaecker) [inline nexus/src/authz/api_resources.rs:60]: Is `_generic` in the function name meant to hammer home the point that all children are treated in the same way? Is there a plan to add non-generic children?

File: nexus/src/authz/mod.rs [hunk 5]
```diff
@@ -2,159 +2,180 @@
...
+//!
+//! Users, roles, and API resources are all stored in the database.  Naturally,
+//! so is the relationship that says a particular user has a particular role for
+//! a particular resource.
+//!
+//! Suppose a built-in user "harry" has the "viewer" role for a Project
+//! "quidditch".  It looks like this:
+//!
+//! ```text
+//! +---> table: "project"
```
Comment (bnaecker) [inline nexus/src/authz/mod.rs:77]: Nice diagram. Did you use a tool to generate this?

File: nexus/src/authz/api_resources.rs (hunk 1 above)
Comment (davepacheco) [reply nexus/src/authz/api_resources.rs]: Fixed in 00ac778eadf087dc0be1d8d652dbeee55521189a.

File: nexus/src/authz/api_resources.rs (hunk 2 above)
Comment (davepacheco) [reply nexus/src/authz/api_resources.rs]: Fixed in 00ac778eadf087dc0be1d8d652dbeee55521189a.

File: nexus/src/authz/api_resources.rs (hunk 3 above)
Comment (davepacheco) [reply nexus/src/authz/api_resources.rs]: Fixed in 00ac778eadf087dc0be1d8d652dbeee55521189a.

File: nexus/src/authz/api_resources.rs (hunk 4 above)
Comment (davepacheco) [reply nexus/src/authz/api_resources.rs:60]: "Yes" and "maybe", respectively.  We'll add more specific children if/when we decide we want to support assigning roles to those resources.  I'm not sure if or when that will happen though.

File: nexus/src/authz/mod.rs (hunk 5 above)
Comment (davepacheco) [reply nexus/src/authz/mod.rs:77]: Thanks!  Nope!  Maybe should have.

File: nexus/src/authz/roles.rs [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+                // grants access to this resource.  In practice, we only use
+                // "parent", and it's clearer to just call this "parent" than
+                // "related_resources_whose_roles_might_grant_access_to_this".)
+                if let Some(parent) = self.parent() {
+                    parent
+                        .fetch_all_related_roles_for_user(
+                            opctx, datastore, authn, roleset,
+                        )
+                        .await?;
+                }
```
Comment (david-crespo) [inline nexus/src/authz/roles.rs:199]: In pulling roles for a `(user, resource)` pair, we're making a query for each level that the resource is nested under the fleet?

File: nexus/src/authz/context.rs [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+        // callers if we didn't return api::external::Error here.
+        let mut roles = RoleSet::new();
+        resource
+            .fetch_all_related_roles_for_user(
+                opctx,
+                &self.datastore,
+                &self.authn,
+                &mut roles,
+            )
+            .await?;
```
Comment (david-crespo) [inline nexus/src/authz/context.rs:87]: This is probably a basic thing about `Box`, but why do you have to use a mutable reference instead of returning `RoleSet` from here? Would a returned `RoleSet` have too short a lifetime?

File: nexus/src/authz/actor.rs [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+use uuid::Uuid;
+
+/// Represents [`authn::Context`] (which is either an authenticated or
+/// unauthenticated actor) for Polar
+#[derive(Clone, Debug)]
+pub struct AnyActor {
+    authenticated: bool,
+    actor_id: Option<Uuid>,
+    roles: RoleSet,
+}
```
Comment (david-crespo) [inline nexus/src/authz/actor.rs:19]: Why is this called `AnyActor` instead of `Actor`?

File: nexus/src/authz/context.rs (hunk 7 above)
Comment (davepacheco) [reply nexus/src/authz/context.rs:87]: Good question.  The reason I had this accept a `&mut RoleSet` rather than returning a `RoleSet` is that we're going to call this function a bunch of times recursively.  These recursive calls can all store their results into the same data structure.  The alternative would involve allocating a new `RoleSet` in each call and then merging it after each child recursive call.  That'd be fine, but it's more code and more work at runtime.

File: nexus/src/authz/actor.rs (hunk 8 above)
Comment (davepacheco) [reply nexus/src/authz/actor.rs:19]: Just to better distinguish it from `AuthenticatedActor`.

(That predates this change -- this code just moved.)

File: nexus/src/authz/actor.rs [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+            roles,
+        }
+    }
+}
+
+impl PartialEq for AnyActor {
+    fn eq(&self, other: &Self) -> bool {
+        self.actor_id == other.actor_id
+    }
+}
```
Comment (david-crespo) [inline nexus/src/authz/actor.rs:36]: When do `AnyActor`s get compared? Is there a risk of creating two actors with the same ID but different `RoleSets` and considering them equal when they're not supposed to be?

File: nexus/src/authz/oso_generic.rs [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+                    Action::Modify => Perm::Modify,
+                    Action::Delete => Perm::Modify,
+                    Action::ListChildren => Perm::ListChildren,
+                    Action::CreateChild => Perm::CreateChild,
+                }
+                .to_string()
+            },
+        )
+    }
+}
```
Comment (david-crespo) [inline nexus/src/authz/oso_generic.rs:90]: Does this basically answer the question "what permission do I need to perform this action?" If so, does that mean only one permission can authorize a particular action, or could this be done in a way that actions map to multiple permissions?

File: nexus/src/authz/actor.rs (hunk 9 above)
Comment (davepacheco) [reply nexus/src/authz/actor.rs]: tl;dr: I think this implementation is vestigial.  I'm removing it in 2ffc5b90.

In theory this _could_ be used in rules in the omicron.polar policy file, like "this actor is allowed to access this resource if they are the built-in 'db-init' user" or "this actor is allowed to access this resource if they are the owner".  I think we'd only ever do this with `AuthenticatedActor` (and we _do_ do that today [here](https://github.com/oxidecomputer/omicron/pull/520/files/9d12b9e7354691aac0139b6b2d7bd06b6742cd85#diff-79878c4356f0c780fbde8574a057daf8129f2205f3a1165d2d6ce5d42e8b148dR44)).

Comment (davepacheco) [reply nexus/src/authz/actor.rs]: Also: in the case that we're comparing two `AuthenticatedActor`s, it doesn't matter what their `RoleSet` is -- we're trying to determine if they're the same person (or machine, or whatever).

File: nexus/src/authz/oso_generic.rs (hunk 10 above)
Comment (david-crespo) [reply nexus/src/authz/oso_generic.rs:90]: In the Oso docs I detect a little ambiguity. Sometimes talk as though actions and permissions are the same, or at least match up one to one. E.g., second argument to `has_permission` is called an action here:

```prolog
allow(actor, action, resource) if
  has_permission(actor, action, resource);
```

https://docs.osohq.com/guides/rbac.html#declare-roles-and-permissions

Comment (davepacheco) [reply nexus/src/authz/oso_generic.rs:90]: > Does this basically answer the question "what permission do I need to perform this action?"

Yes.

> In the Oso docs I detect a little ambiguity. Sometimes they talk as though actions and permissions are the same, or at least match up one to one.

Yeah, I can't say I understand the intended distinction between actions and permissions.  As you said: the thing seems to come in as an "action" and be used as a "permission", and all the examples I've seen are strings.

I wrote this (tiny) abstraction layer to try to make that a bit more explicit.  Nexus at-large authorizes you for an _action_ and we translate it (mostly 1-1) to a _permission_ and use permissions in the policy.

> If so, does that mean only one permission can authorize a particular action, or could this be done in a way that actions map to multiple permissions?

I think the way this is set up today, that's true.  In principle, we could have `to_perm()` return a list of permissions and the policy could check all of them separately.  In practice, I think that sort of thing is generally discouraged.

Comment (david-crespo) [reply nexus/src/authz/oso_generic.rs:90]: I see, that makes sense. Could you get rid of Perm entirely then and impl `Display` directly on `Action` instead of converting `Action` to `Perm` and then calling `to_string` on the latter?

Related, why is there a delete action but no delete permission?

Comment (david-crespo) [reply nexus/src/authz/oso_generic.rs:90]: I see now they are called `permissions` in the policy, so I take it the `Perm` thing is meant to make clear that what we're calling Actions elsewhere become permissions once we're inside the policy.

Comment (davepacheco) [reply nexus/src/authz/oso_generic.rs:90]: > Could you get rid of Perm entirely then and impl Display directly on Action instead of converting Action to Perm and then calling to_string on the latter?

I think we could, but I think that conflates things (the display name of an Action does not need to be exactly the same string that we use all over the policy file).  It feels like it makes it easier to confuse an arbitrary String with a Permission.

> Related, why is there a delete action but no delete permission?

Short answer: because there's no role in the policy that would have "delete" and not "modify", so there's no point in distinguishing them in the policy.  (It _is_ useful to distinguish the actions, and that's the one case where actions don't map 1-1 with permissions.)

Basically I wrote the existing policy (i.e., the contents of omicron.polar) based on concrete use cases and I didn't see an obvious use case for someone being able to delete a thing but not be able to modify it.

Comment (davepacheco) [reply nexus/src/authz/oso_generic.rs:90]: > I see now they are called `permissions` in the policy, so I take it the `Perm` thing is meant to make clear that what we're calling Actions elsewhere become permissions once we're inside the policy.

Yeah that's a much better way to put it than what I said!

File: nexus/src/authz/roles.rs [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+                //
+                // (In general, there could be another resource with _any_ kind
+                // of relationship to this one that grants them a role that
+                // grants access to this resource.  In practice, we only use
+                // "parent", and it's clearer to just call this "parent" than
+                // "related_resources_whose_roles_might_grant_access_to_this".)
+                if let Some(parent) = self.parent() {
+                    parent
+                        .fetch_all_related_roles_for_user(
+                            opctx, datastore, authn, roleset,
```
Comment (david-crespo) [inline nexus/src/authz/roles.rs:196]: Doesn't need to be addressed in this PR, but since we have the whole chain of `(ID, resource type)` in memory here, we should be able to pull them out into a list and have `role_asgn_builtin_list_for` query for them all at once in a single query.

**Review by david-crespo — APPROVED** [review]
exciting and scary

File: nexus/src/authz/oso_generic.rs (hunk 10 above)
Comment (davepacheco) [reply nexus/src/authz/oso_generic.rs:79]: > I wouldn't mind a line explaining this in the Perm doc comment.

I've done this in 7020feba03be14f016bea5e629c956c8c9096c34.  I also moved the guts of the conversion out of the "to_perm" Polar method and into an explicit `impl From<&Action> for Perm`.  To me, that more clearly communicates that `Perm`'s job is to aid in this translation.

File: nexus/src/authz/roles.rs (hunk 11 above)
Comment (davepacheco) [reply nexus/src/authz/roles.rs:196]: Yeah, it might be a nice improvement if this function, rather than making the query and updating `RoleSet`, actually just stored the relevant (resource_type, id) tuples into a BTreeSet, and then we used that at the end to build a sort of monstrous query in `role_asgn_builtin_list_for`.
